### DISCUSSION TITLE
How to pop nedit-ng to the front, when opening an additional file, on Mac [FIXED]

### DISCUSSION THREAD

**Discussion by grege2** [eteran/nedit-ng#207] [General]
I can see some discussion back in the Issues, about whether nedit-ng or nc-ng pops to the front when opened, or what "detached tabs" should do, etc. But I thought I'd open a new discussion since the Mac build is changing a bit recently.

So, in nedit-X/Motif, any new file added to the nedit "open files" seemed to pop the whole window to the front. But with nedit-ng it just stays put, sometimes hidden behind other windows. No confirmation that anything happened at all in fact, a bit strange. This is when say using "nc-ng firstfile", and opening an additional file with "nc-ng newfile".

I've tried to manually attach a "bring window to the front" action purely in some kind of Mac magic, but no luck so far.

Would nedit-ng be able to raise itself to the front of the window stack, when a new tab is opened and populated?

(I have a nagging feeling that somewhere back in the Issues there's someone who specifically -doesn't- want that, you just never know what folks want).

Anyway, comments welcome.

Ciao, GE

**Comment by eteran** [comment]
IIRC, the intended behavior is definitely to bring the window to the front. But it has proven difficult to get quite right. I'll definitely continue to investigate this, as the few issues regarding window placement, size, and focus when using nc-ng is pretty much the only outstanding issues that I consider regressions from nedit classic.

**Comment by grege2** [comment]
Hi, I came up with an AppleScript solution to my needs. I invoke nedit-ng with a script `nc`
```
#!/bin/bash

# Interlude to raise any existing nedit-ng process to the front, 
#  then open new tab(s) with required file(s)

# Check manually if nedit-ng is running, more predictable than osascript's behaviour
ps -ejww | grep nedit-ng |  grep -v grep >/dev/null 2>&1
STATUS=$?

if [ $STATUS -eq 0 ] ; then 
  osascript ~/binhi/pop-nedit &        # Raise window
  if [ "$*" == "" ] ; then exit ; fi   # Nothing more to do, no new files to be opened
fi

set -x    # Show what's finally happening
~/binhi/nc-ng -geometry 140x36   $*
```
and the AppleScript script `pop-nedit` is ..
```
tell application "/Users/gedwards/binhi/nedit-ng"
  activate
end tell
```
It seems to work well for my fairly narrow needs:
 - open the first file with `nc file1 [file2 ..]`
 - (the nedit window may then get partly or fully hidden by some activity)
 - open further files with `nc file3 [file4 .. ]` - the nedit window is raised to the front and gets focus
 - to just raise the window, a plain `nc` is ok
 - a plain `nc` with nothing already running, opens an empty `Untitled` tab as per usual
 - the `nc` also restores and raises and focuses a minimised window, ie. in the Dock.

I'll give it more testing and see how it goes. I haven't considered things like detached tabs, running with more than one nedit-ng window open, opening files from the File menu, subtleties of nedit-ng in "client" vs. "server" mode.

Edit: Of course you can raise a nedit-ng window by clicking on the icon in the Dock(but .. only if the Dock is visible). Also with Cmd-Tab to cycle through active apps. Also F9 to show all windows. My nedit habits are a bit locked in to the command line.

Ciao, GE.

**Comment by eteran** [comment]
@grege2 this MAY be fixed on the current master. could you give it a try and let me know?

**Comment by grege2** [comment]
Hi, I've given it a brief testing and looks good. Any use of "nc-ng" from the command line raises an existing window and does the right thing, ie. 
 - activates the correct tab, or an Untitled tab if no file given
 - tab is active and has focus
 - window also restores from Dock

Also works ok with "detached tabs", the correct window is raised, containing the necessary tab. I don't work this way much, I'm not a good tester for that. I don't have a dual-screen with several edit windows open.

When there are several windows open, from detaching tabs, there's a choice for where to open a new file. It seems to go in the most recently created window. Also some random closing of windows doesn't break things, nedit knows it doesn't have "tab n = file xxx" any more and opens and raises a new tab fine.

Also it's much quicker than my AppleScript interlude. That was quick enough, maybe 1/4 of a second, a little variable, not a problem, but juuuust long enough for a few brain cells to wonder "is this thing going to work?", but native nc-ng is now immediate.

So, looking great, thanks. I should use it in anger for a couple of days to shake out anything else.

(And, why am I playing with my coding hobbies on Xmas Day ???  Oh well .. I do have a beer and a pile of Xmas lunch leftovers .. ) ( beer's gone, I'll get the Shiraz ..)
